draft-ietf-taps-interface.md

```diff
@@ -1163,53 +1212,94 @@ it is bound to the Preconnection during the pre-establishment phase:
 ~~~
 Preconnection.FrameWith(Framer)
 
-OctetArray := Framer.Frame(Message)
+OctetArray := Framer.Frame(messageData)
 ~~~
 
 Sender-side framing is a convenience feature of the interface, for parity with
 receiver-side framing (see {{receive-framing}}).
 
 # Receiving Data {#receiving}
 
-Once a Connection is established, Messages may be received on it. The application can indicate that it is ready to receive Messages by calling Receive() on the Connection.
+Once a Connection is established, it can be used for receiving data. As with
+sending, data is received in terms of Messages. Receiving is an asynchronous
+operation, in which each call to Receive enqueues a request to receive new
+data from the connection. Once data has been received, or an error is encountered,
+an event will be delivered to complete the Receive request (see {{receive-events}}).
+
+As with sending, the type of the Message to be passed is dependent on the
+implementation, and on the constraints on the Protocol Stacks implied by the
+Connection's transport parameters.
+
+## Enqueuing Receives
+
+Receive takes two parameters to specify the length of data that an application
+is willing to receive, both of which are optional and have default values if not
+specified. 
 
 ~~~
-Connection.Receive(ReceiveHandler, maxLength)
+Connection.Receive(minIncompleteLength, maxLength)
 ~~~
 
-Receive takes a ReceiveHandler, which can handle the Received Event and the
-ReceiveError error. Each call to Receive will result in at most one Received
-event being sent to the handler, though implementations may provide
-convenience functions to indicate readiness to receive a larger but finite
-number of Messages with a single call. This allows an application to provide
-backpressure to the transport stack when it is temporarily not ready to
-receive messages.
+By default, Receive will try to deliver complete Messages in a single event ({{receive-complete}}).
 
-Receive also takes an optional maxLength argument, the maximum size (in bytes
-of data) Message the application is currently prepared to receive. The default
+The application can set a minIncompleteLength value to indicates the smallest partial 
+Message data size in bytes that should be delivered in response to this Receive. By default, 
+this value is infinite, which means that only complete Messages should be delivered. 
+If this value is set to some smaller value, the associated recieve event will be triggered 
+only when at least that many bytes are available, or the Message is complete with fewer
+bytes, or the system needs to free up memory. Applications should always
+check the length of the data delivered to the receive event and not assume
+it will be as long as minIncompleteLength in the case of shorter complete Messages
+or memory issues.
+
```

Comment:
I understand that the applications should derive the length of the delivered data from the messageData. Perhaps this should be made more explicit? Before I got to the ReceivedPartial Event when reading, where the messageContext was reused for multiple events, I was thinking that the length of the data delivered may be provided to the application as part of the messageContext meta data.

Reply:
Good point, I'll add an explicit comment about this.